ncm-afsclt/src/main/perl/afsclt.pm

```diff
@@ -2,997 +2,314 @@
-    LC::Check::file("CellServDB",
-		    "contents"    => $cellservdb_content,
-		    "destination" => $localcelldb,
-		    "owner"       => 0,
-		    "mode"        => 0444,
-		   );
-    $self->update_afs_cells();
-  }
-
-  return 1;
 }
 
-# This code comes from sue/feature/afs/update.pl, slightly modified
-# to work in our context (replace afsutil with "fs")
-#
 # update the list of known AFS cells and run "fs newcell" when needed
 # This will only add new cells, not remove currently known ones.
 #
 sub update_afs_cells ( $$ ) {
-    my($self) = @_;
-    my($cell, %seen, %ipaddrs, @hosts, $todo, @todo, $error, $afsutil);
-    local(*FH, $_);
-    # init
-
-    # read CellServDB
-    # take it from the real file
-    if( ! open(*FH, $localcelldb)) {
-      $self->warn("cannot read local $localcelldb, giving up");
-      return 0;
-    }
+    my ($self) = @_;
+    my ( $cell, %seen, %ipaddrs, @hosts, $todo, @todo, $error, $afsutil );
 
-    while (<FH>) {
+    my $localcelldb_fh = CAF::FileEditor->new( $localcelldb, log => $self );
+    for ( split ( /\n/, "$localcelldb_fh" ) ) {
         if (/^>(.\S+)/) {
-            # new cell
             $cell = $1;
             $ipaddrs{$cell} = [];
-        } elsif (/^(\S+)\s+\#\s*(\S+)/) {
+        }
+        elsif (/^(\S+)\s+\#\s*(\S+)/) {
+
             # new entry
-            push(@{$ipaddrs{$cell}}, $1);
+            push ( @{ $ipaddrs{$cell} }, $1 );
         }
     }
-    close(FH);
-    foreach $cell (keys(%ipaddrs)) {
-        @{$ipaddrs{$cell}} = sort(@{$ipaddrs{$cell}});
-        $self->Debug("CellServDB $cell -> @{$ipaddrs{$cell}}");
+    $localcelldb_fh->close();
+    foreach $cell ( keys(%ipaddrs) ) {
+        @{ $ipaddrs{$cell} } = sort( @{ $ipaddrs{$cell} } );
+        $self->debug( 1, "CellServDB $cell -> @{$ipaddrs{$cell}}" );
     }
+
     # read known cells
     @todo = ();
-    my @current = `fs listcell -numeric 2>&1`;
+    my $proc = CAF::Process->new( [ "fs", "listcell", "-numeric" ], log => $self, keeps_state => 1 );
 
-    if($? >> 8 ) {
-      $self->info("cannot read current AFS cell info via fs:\n".join('',@current));
-      return 1;
+    if ( $? >> 8 ) {
+        $self->error( "Cannot read current AFS cell info via fs:\n" . $proc->output() );
+        return 1;
     }
-    foreach (@current) {
+
+    foreach ( split ( /\n/, $proc->output() ) ) {
         chomp($_);
         next unless /^Cell\s+(\S+)\s+on hosts\s+(\S.+)\.$/;
         $cell = $1;
-        @hosts = split(/\s+/, $2);
+        @hosts = split ( /\s+/, $2 );
         $seen{$cell} = 1;
-        $self->Debug("fs listcell: $cell -> @hosts");
-        next unless $ipaddrs{$cell} && @{$ipaddrs{$cell}};
+        next unless $ipaddrs{$cell} && @{ $ipaddrs{$cell} };
         @hosts = sort(@hosts);
-        if ("@hosts" ne "@{$ipaddrs{$cell}}") {
-            push(@todo, $cell);
-            $self->Debug("cell info for $cell changed from @hosts to @{$ipaddrs{$cell}}");
+        if ( "@hosts" ne "@{$ipaddrs{$cell}}" ) {
```

Comment:
no it seems to do space separated join, so for IP addresses this is safe.
```
# perl -e '@hosts=qw(a b c);$b={x => [qw(abc)]}; print "@hosts @{$b->{x}}\n"'
a b c abc
```
learned something new :wink: